### PR TITLE
client_channel, client_auth: rewrite disallowed status codes from the control plane

### PR DESCRIPTION
**Pull request by markdroth**
This implements [gRFC A54](https://github.com/grpc/proposal/blob/master/A54-restrict-control-plane-status-codes.md).

### CODE REVIEW

File: src/core/lib/security/transport/client_auth_filter.cc [hunk 1]
```diff
@@ -157,7 +160,15 @@ ArenaPromise<absl::StatusOr<CallArgs>> ClientAuthFilter::GetCallCredsMetadata(
 
   auto client_initial_metadata = std::move(call_args.client_initial_metadata);
   return TrySeq(
-      creds->GetRequestMetadata(std::move(client_initial_metadata), &args_),
+      Seq(creds->GetRequestMetadata(std::move(client_initial_metadata), &args_),
```
Comment (yashykt) [inline src/core/lib/security/transport/client_auth_filter.cc:163]: I see, so we need to check all filters where a per-channel logic is affecting an RPC's status. Have you checked all filters? 
Is there a place where we can document this for new filter writers?

**Comment by yashykt** [comment]
I'll leave it to @ctiller to review the filter changes

**Review by yashykt — APPROVED** [review]

File: src/core/lib/security/transport/client_auth_filter.cc (hunk 1 above)
Comment (markdroth) [reply src/core/lib/security/transport/client_auth_filter.cc:163]: The goal here is not to check all filters.  The [gRFC](https://github.com/grpc/proposal/blob/master/A54-restrict-control-plane-status-codes.md#proposal) specifies 4 individual cases that need to be covered, and those are the ones covered in this PR.

File: src/core/lib/channel/status_util.cc [hunk 2]
```diff
@@ -109,3 +111,28 @@ bool grpc_status_code_from_int(int status_int, grpc_status_code* status) {
...
+    // https://github.com/grpc/proposal/blob/master/A54-restrict-control-plane-status-codes.md.
+    case absl::StatusCode::kInvalidArgument:
+    case absl::StatusCode::kNotFound:
+    case absl::StatusCode::kAlreadyExists:
+    case absl::StatusCode::kFailedPrecondition:
+    case absl::StatusCode::kAborted:
+    case absl::StatusCode::kOutOfRange:
+    case absl::StatusCode::kDataLoss: {
+      return absl::InternalError(
+          absl::StrCat("illegal status code from ", source,
```
Comment (ctiller) [inline src/core/lib/channel/status_util.cc:130]: nit: capitalize `I` in illegal

File: src/core/lib/channel/status_util.cc [hunk 3]
```diff
@@ -109,3 +111,28 @@ bool grpc_status_code_from_int(int status_int, grpc_status_code* status) {
   *status = static_cast<grpc_status_code>(status_int);
   return true;
 }
+
+namespace grpc_core {
+
+absl::Status MaybeRewriteIllegalStatusCode(const absl::Status& status,
```
Comment (ctiller) [inline src/core/lib/channel/status_util.cc:117]: If we changed this signature to:

```
absl::Status MaybeRewriteIllegalStatusCode(absl::Status status, absl::string_view source)
```
we could std::move the status into this function in the majority of (all?) cases, and then use the RVO to make this rewriting essentially free even if there's an error message attached to a valid status.

**Review by ctiller — APPROVED** [review]
It's probably worth addressing the performance comment I made

File: src/core/lib/channel/status_util.cc (hunk 3 above)
Comment (markdroth) [reply src/core/lib/channel/status_util.cc:117]: Done.

File: src/core/lib/channel/status_util.cc (hunk 2 above)
Comment (markdroth) [reply src/core/lib/channel/status_util.cc:130]: Done.

**Pull request merged**